# Buchhaltung mit Excel und Numpy (Übung)


Open in Colab: [![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/febse/opt2025-de/blob/main/Lineare-Algebra.ipynb)


- [Konten](https://github.com/febse/data/raw/refs/heads/main/opt/Konten.xlsx)
- [Konten (shared)](https://febunisofia-my.sharepoint.com/:x:/g/personal/amarov_feb_uni-sofia_bg/EQNsytBLVrdHjWlTNNh7w7AB5irt4Xxa-TEYyeyVdNM7sQ?e=hysC4M)

Stellen wir uns vor, dass wir mehrere Bankkonten haben, z.B. ein Bankkonto in BGN in Bulgarien, ein PayPal-Konto in USD und ein Revolut-Konto in Euro. Wir möchten wissen, wie viel Geld wir insgesamt haben. Die einzelne Kontostände seien z.B. $x = 1000$ BGN in dem BGN-Konto, $y = 500$ USD in dem PayPal-Konto und $z = 200$ Euro in dem Revolut-Konto. Wir können diese Kontostände nicht direkt addieren, weil sie in verschiedenen Währungen sind. Um unser Gesamtvermögen zu berechnen, müssen wir alle Kontostände in dieselbe Währung umrechnen. Wir können das mit den Wechselkursen machen:

- 1 BGN = 0.51 Euro
- 1 USD = 0.91 Euro
- 1 Euro = 1 Euro

Unser Gesamtkontostand in Euro ist dann:

$$
0.51 \cdot 1000 + 0.91 \cdot 500 + 1 \cdot 200 = 510 + 455 + 200 = 1165 \text{ Euro}
$$

Allgemeiner können wir unser Gesamtkontostand in Euro mit einer Funktion $F$ berechnen:

$$
F(x, y, z) = 0.51 x + 0.91 y + 1 z
$$

Was diese Funktion tut ist einfach zuerst die Messeinheiten ihrer Inputs zu ändern (BGN in Euro, USD in Euro und Euro in Euro), dann zählt sie alles zusammen. Wir können sie z.B. mit dem (langen) Namen Gesamtkontostandineurorechner nennen.



## Vereinfachung

Schauen wir uns den Gesamtkontostandineurorechner allgemeiner an.

$$
F(x, y, z) = a x + b y + c z
$$

Da die Funktion $F$ immer ein und dasselbe tut (umrechnen, dann addieren), brauchen wir nur ihre Parameter (Umrechnungsfaktoren) $a$, $b$ und $c$ zu kennen. In unserem Beispiel sind die Parameter $a = 0.51$, $b = 0.91$ und $c = 1$. 

$$
[0.51, 0.91, 1]
$$

Wir können die Funktion und ihre Parameter (Wechselkurse) [$0.51$, $0.91$, $1$] als eine **Operation** auffassen: umrechnen, addieren. Die Daten, die in diese Operation einfließen, sind die Kontostände [$1000$, $500$, $200$]. Um das ganze etwas übersichtlicher zu machen, können wir die Kontostände und die Wechselkurse in zwei Listen packen:


::: {#tbl-panel layout-ncol=2}
| Bank (BGN) | PayPal (USD) | Revolut (EUR) |
|------------|--------------|---------------|
| 1000       | 500          | 200           |

: Kontostände (Daten) {#tbl-accounts-1}

| Alles (EUR)                   |
|-------------------------------|
| 0.51                          |
| 0.91                          |
| 1                             |
|-------------------------------|
| 1165 EUR                      |


: Operationen {#tbl-operations-1}

Kontostände und Wechselkurse (eine Operation)

:::


Nun stellen wir uns vor, dass wir nicht nur den Gesamtwert in Euro berechnen wollen, sondern auch noch die folgenden Zusammenfassungen:

- den Gesamtwert in Euro
- den Wert in Euro nur des PayPal-Kontos
- den Wert in USD des PayPal-Kontos
- den Durchschnittswert in Euro
- den Gesamtwert vom bulgarischen Konto und dem PayPal-Konto (in Euro)

Mit unserer Darstellung der Daten und Operationen geht das sehr einfach. Wir müssen nur herausfinden, wie diese Operationen als Parameter aussehen.

<!-- - Der Gesamtwert in BGN: das haben wir schon: [$0.51$, $0.91$, $1$]
- Der Gesamtwert in Euro des PayPal-Kontos: [$0$, $0.91$, $0$]
- Der Wert in USD des PayPal-Kontos: [$0$, $1$, $0$]
- Der Durchschnittswert in Euro: [$0.51 / 3$, $0.91 / 3$, $1 / 3$]
- Der Gesamtwert vom Bankkonto und vom Revolut-Konto (in Euro): [$0.51$, $0$, $1$]
- Der Gesamtwert vom Bankkonto und vom PayPal-Konto abzüglich des Revolut-Kontos (in Euro): [$0.51$, $0.91$, $-1$] -->

Sie können sich das vorstellen, als ob wir ein und dasselbe Wasser durch fünf verschiedene Siebe gießen. Jedes Sieb hat eine andere Form, aber sie alle haben die gleiche Funktion: sie filtern das Wasser. Das Wasser ist in unserem Fall die Liste der Kontostände und die Siebe sind die Parameter, die wir oben beschrieben haben.

Nun brauchen wir diese Operationen nur noch in unsere Liste einzufügen. Fügen wir die ersten zwei hinzu:

Jetzt haben wir den Gesamtkontostand in Euro und den Kontostand in Euro des PayPal-Kontos.

Fügen wir nun alle Operationen hinzu:

::: {#tbl-panel layout-ncol=2}
| Bank (BGN) | PayPal (USD) | Revolut (EUR) |
|------------|--------------|---------------|
| 1000       | 500          | 200           |

: Kontostände (Daten) {#tbl-accounts-2}

| Alles (EUR)                   | Nur PayPal (EUR)                    |
|-------------------------------|-------------------------------------|
| 0.51                          | 0                                   |
| 0.91                          | 0.91                                |
| 1                             | 0                                   |
|-------------------------------|-------------------------------------|
| 1165 EUR                      | 455 EUR                             |


: Operationen {#tbl-operations-2}

Kontostände und Wechselkurse (zwei Operationen)

:::


In [ ]:
# 




::: {#tbl-panel layout-ncol=2}
| Bank (BGN) | PayPal (USD) | Revolut (EUR) |
|------------|--------------|---------------|
| 1000       | 500          | 200           |

: Kontostände (Daten) {#tbl-accounts-3}

| Alles (EUR)                   | Nur PayPal (EUR)  | Nur PayPal (USD) | Durchschnitt (EUR) |
|-------------------------------|-------------------|------------------|--------------------|
| 0.51                          | 0                 | 0                | 0.51/3             |
| 0.91                          | 0.91              | 1                | 0.91/3             |
| 1                             | 0                 | 0                | 1 /3               |
|-------------------------------|-------------------|------------------|--------------------|
| 1165 EUR                      |   455 EUR         | 500 USD          | 388.3 EUR          |


: Operationen {#tbl-operations-3}

Kontostände und Wechselkurse (vier Operationen)

:::



## Mehr als eine Person (mehr Daten)

Was passiert, wenn wir zwei Personen haben, die Information über ihre drei Konten zusammenfassen wollen? Für unsere Tabellen bedeutet das einfach noch eine Zeile mit Daten. Diese zweite Zeile fließt durch dieselben Siebe wie die erste. Die Ergebnisse der Operationen schreiben wir einfach unter die Ergebnisse für die erste Zeile. Weil die Tabellen unpraktisch breit werden, zeigt das @tbl-operations-persons nur für die ersten zwei Operationen.


::: {#tbl-panel layout-ncol=2}
|Person | Bank (BGN) | PayPal (USD) | Revolut (EUR) |
|-------|------------|--------------|---------------|
|Boyko  | 1000       | 500          | 200           |
|Nikoletta| 800      | 2000         | 100           |

: Kontostände (Daten) {#tbl-accounts-persons}

|Person   | Alles (EUR)                   | Nur PayPal (EUR)                    |
|---------|-------------------------------|-------------------------------------|
|         | 0.51                          | 0                                   |
|         | 0.91                          | 0.91                                |
|         | 1                             | 0                                   |
|         |-------------------------------|-------------------------------------|
|Boyko    | 1165 EUR                      | 455 EUR                             |
|Nikoletta| 2328 EUR                      | 1820 EUR                            |


: Operationen {#tbl-operations-persons}

Kontostände und Wechselkurse (zwei Operationen, zwei Personen)
:::


Da die Tabellen etwas umständlich werden, wenn wir weitere Operationen und Daten hinzufügen, können wir sie etwas kompakter aufschreiben.



## Eine Person, eine Operation

@tbl-accounts-1 und @tbl-operations-1 können wir kompakter auch so schreiben:

$$
\begin{pmatrix}
1000 & 500 & 200 \\
\end{pmatrix}
\begin{pmatrix}
0.51 \\
0.91 \\
1 \\
\end{pmatrix}
= \begin{pmatrix}
1165 \\
\end{pmatrix}
$$

In [2]:
import numpy as np

balances_boyko = np.array([1000, 500, 200])
alles_in_euro = np.array([0.51, 0.91, 1])

# Per Hand
print("Alles in Euro: ", 1000 * 0.51 + 500 * 0.91 + 200 * 1)

# Schneller und bequemer mit NumPy
# Wir berechnen eigentlich das Skalarprodukt der beiden Vektoren (auf Englisch: dot product)

balances_boyko.dot(alles_in_euro)


Alles in Euro:  1165.0


np.float64(1165.0)

In [3]:
# Oder auch so

balances_boyko @ alles_in_euro

np.float64(1165.0)

## Eine Person, mehrere Operationen

@tbl-accounts-2 und @tbl-operations-2 können wir auch so schreiben:

$$
\begin{align*}
\begin{pmatrix}
1000 & 500 & 200 \\
\end{pmatrix}
&
\begin{pmatrix}
0.51 & 0 & 0 & 0.51 / 3 \\
0.91 & 0.91 & 1 & 0.91 / 3 \\
1 & 0 & 0 & 1 / 3\\
\end{pmatrix} \\
= &
\begin{pmatrix}
1165.0 & 455.0 & 500.0 & 388.3 \\
\end{pmatrix}
\end{align*}
$$

In [4]:
# Wir definieren die einzelnen Operationen
nur_paypal_euro = np.array([0, 0.91, 0])
nur_paypal_usd = np.array([0, 1, 0])
durchschnitt_euro = np.array([0.51 / 3, 0.91 / 3, 1 / 3])
bank_und_revolut_euro = np.array([0.51, 0, 1])

# Wir stapeln die einzelnen Operationen _nebeneinander_ (axis = 1) in eine Matrix
operations4 = np.stack([
    alles_in_euro,
    nur_paypal_euro,
    nur_paypal_usd,
    durchschnitt_euro,
], axis=1)

# Schauen wir uns an, was wir konstruiert haben
operations4

array([[0.51      , 0.        , 0.        , 0.17      ],
       [0.91      , 0.91      , 1.        , 0.30333333],
       [1.        , 0.        , 0.        , 0.33333333]])

In [5]:
operations4.shape

(3, 4)

In [6]:
balances_boyko @ operations4

array([1165.        ,  455.        ,  500.        ,  388.33333333])

## Zwei Personen, eine Operation

$$
\underset{\text{Daten}}{\begin{pmatrix}
1000 & 500 & 200 \\
800 & 2000 & 100 \\
\end{pmatrix}}
\underset{\text{Operation}}{\begin{pmatrix}
0.51 \\
0.91 \\
1 \\
\end{pmatrix}}
= \underset{\text{Neue Daten}}{\begin{pmatrix}
1165 & 2328 \\
\end{pmatrix}}
$$

In [7]:
balances_nikoletta = np.array([800, 2000, 100])

# Wir stapeln die einzelnen Operationen _untereinander_ (axis = 0) in eine Matrix
balances_all = np.stack([balances_boyko, balances_nikoletta])
balances_all

array([[1000,  500,  200],
       [ 800, 2000,  100]])

In [8]:
balances_all @ alles_in_euro

array([1165., 2328.])

# Zwei Personen, zwei Operationen

$$
\underset{\text{Daten}}{\begin{pmatrix}
1000 & 500 & 200 \\
800 & 2000 & 100 \\
\end{pmatrix}}
\underset{\text{Operation}}{\begin{pmatrix}
0.51 & 0.91 & 1 \\
0 & 0.91 & 0 \\
\end{pmatrix}}
= \underset{\text{Neue Daten}}{\begin{pmatrix}
1165 & 455 \\
2328 & 1820 \\
\end{pmatrix}}
$$

In [9]:
operations2 = np.stack([
    alles_in_euro,
    nur_paypal_euro,
], axis=1)

operations2

array([[0.51, 0.  ],
       [0.91, 0.91],
       [1.  , 0.  ]])

In [10]:
balances_all @ operations2

array([[1165.,  455.],
       [2328., 1820.]])

# Zwei Personen, mehrere Operationen

Hier schreiben wir die Daten aus @tbl-accounts-persons und die Operationen aus @tbl-operations-persons kompakter auf:

$$
\begin{align*}
\begin{pmatrix}
1000 & 500 & 200 \\
800 & 2000 & 100 \\
\end{pmatrix}
\begin{pmatrix}
0.51 & 0 & 0 & 0.51 / 3 \\
0.91 & 0.91 & 1 & 0.91 / 3 \\
1 & 0 & 0 & 1 / 3\\
\end{pmatrix} 
= 
\begin{pmatrix}
1165.0 & 455.0 & 500.0 & 388.3 \\
2328.0 &  1820.0 &  2000.0 &  776.0 \\        
\end{pmatrix}
\end{align*}
$$


In [11]:
balances_all @ operations4

array([[1165.        ,  455.        ,  500.        ,  388.33333333],
       [2328.        , 1820.        , 2000.        ,  776.        ]])

## Vertauschung der Operationen

### Eine Person, eine Operation

Wir können die Reihenfolge der Operationen vertauschen und die erste Tabelle als Operationen auffassen, die zweite als die Daten.

$$
\begin{pmatrix}
0.51 & 0.91 & 1 \\
\end{pmatrix}
\begin{pmatrix}
1000 \\ 500 \\ 200 \\
\end{pmatrix}
= 
0.51 \cdot 1000 + 0.91 \cdot 500 + 1 \cdot 200 = 510 + 455 + 200 = 1165 \text{ Euro}
$$


In [12]:
alles_in_euro @ balances_boyko

np.float64(1165.0)

## Zwei Operationen, eine Person

$$
\underset{\text{Operationen}}{\begin{pmatrix}
0.51 & 0.91 & 1 \\
0 & 0.91 & 1 \\
\end{pmatrix}}
\underset{\text{Daten}}{\begin{pmatrix}
1000 \\ 500 \\ 200 \\
\end{pmatrix}}=
\underset{\text{Neue Daten}}{\begin{pmatrix}
1165 \\ 455.0 \\
\end{pmatrix}}
$$


In [13]:
operations2.T @ balances_boyko

array([1165.,  455.])

In [14]:
balances_all.T

array([[1000,  800],
       [ 500, 2000],
       [ 200,  100]])

In [15]:
operations2

array([[0.51, 0.  ],
       [0.91, 0.91],
       [1.  , 0.  ]])

In [16]:
operations2.T

array([[0.51, 0.91, 1.  ],
       [0.  , 0.91, 0.  ]])

### Mehrere Operationen, mehrere Personen

$$
\underset{\text{Operationen}}{\begin{pmatrix}
0.51 & 0.91 & 1 \\
0 & 0.91 & 0 \\
0 & 1 & 0 \\
0.51 / 3 & 0.91 / 3 & 1 / 3 \\
\end{pmatrix}}
\underset{\text{Daten}}{\begin{pmatrix}
1000 & 800 &  \\
500 & 2000 & \\
200 & 100 &
\end{pmatrix}}
=
\underset{\text{Neue Daten}}{\begin{pmatrix}
1165 & 2328 \\
455 & 1820 \\
500 & 2000 \\
388.3 & 776 \\
\end{pmatrix}}
$$


In [17]:
operations4.T

array([[0.51      , 0.91      , 1.        ],
       [0.        , 0.91      , 0.        ],
       [0.        , 1.        , 0.        ],
       [0.17      , 0.30333333, 0.33333333]])

In [18]:
balances_all.T

array([[1000,  800],
       [ 500, 2000],
       [ 200,  100]])

In [19]:
operations4.T @ balances_all.T

array([[1165.        , 2328.        ],
       [ 455.        , 1820.        ],
       [ 500.        , 2000.        ],
       [ 388.33333333,  776.        ]])

In der Schule haben Sie Algebra gelernt, die sich mit Zahlen beschäftigt. Das Wort Algebra stammt von dem arabischen Wort "al-jabr" ab, das ursprünglich "Reparatur" oder "Wiederherstellung" bedeutete.

Lineare Algebra untersucht die Eigenschaften und Beziehungen zwischen einfachen und leicht verständliche Transformationen.

Als Beispiel nehmen wir eine Funktion $f(x) = 0.51 x$, die als Input eine Zahl $x$ (Preis in BGN) erhält und als Ausgabe eine neue Zahl (Preis in Euro) zurückgibt. $0.51 \approx 1 / 1.95583$ ist der Wechselkurs zwischen BGN und Euro. Das ist ein Beispiel für eine lineare Funktion (Abbildung) und sie had die bequeme Eigenschaft, dass wenn wir den Input verdoppeln (also $x$ verdoppeln), dann verdoppelt sich auch der Output (also $f(x)$ verdoppelt sich).

$$
\begin{align*}
f(1) & = 0.51 \cdot 1 = 0.51 \\
f(2) & = 0.51 \cdot 2 = 1.02 \\
f(4) & = 0.51 \cdot 4 = 2.04 \\
\end{align*}
$$

oder im allgemeinen:

$$
f(a x) = a f(x)
$$

Andere Funktionen wie z.B. $g(x) = x^2$ haben diese Eigenschaft nicht:

$$
\begin{align*}
g(1) & = 1^2 = 1 \\
g(2) & = 2^2 = 4 \neq 2 g(1) \\
\end{align*}
$$

Sogar die Funktion $h(x) = x + 2$, die 2 zu ihrem Input addiert, hat diese Eigenschaft nicht:

$$
\begin{align*}
h(1) & = 1 + 2 = 3 \\
h(2) & = 2 + 2 = 4 \neq 2 h(1) \\
\end{align*}
$$

Eine andere Eigenschaft von linearen Funktionen wie $f(x)$ ist, dass wenn wir die Funktion auf die Summe von zwei Zahlen anwenden, dann ist das Ergebnis gleich der Summe der Ergebnisse der Funktion auf die einzelnen Zahlen:

$$
\begin{align*}
f(1 + 2) & = f(3) = 0.51 \cdot 3 = 1.53 \\
f(1) + f(2) & = 0.51 + 0.51 \cdot 2 = 1.53 \\
\end{align*}
$$

Im allgemeinen:

$$
f(x + y) = f(x) + f(y)
$$


## Lineare Funktionen
(XXX, TODO add)

Ist das eine lineare Funktion? Ja, weil sie die zwei Eigenschaften hat, die wir oben beschrieben haben:

Verdoppeln wir die Inputs, verdoppelt sich auch der Output:

$$
F(2x, 2y, 2z) = 0.51 \cdot (2x) + 0.91 \cdot (2y) + 1 \cdot (2z) = 2 \cdot (0.51 \cdot x + 0.91 \cdot y + 1 \cdot z) = 2 F(x, y, z)
$$

Wenn wir die Inputs addieren, addiert sich auch der Output:

$$
\begin{align*}
F(x + 2, y + 3, z + 5) & = 0.51 \cdot (x + 2) + 0.91 \cdot (y + 3) + 1 \cdot (z + 5) \\
& = 0.51 \cdot x + 0.51 \cdot 2 + 0.91 \cdot y + 0.91 \cdot 3 + 1 \cdot z + 1 \cdot 5 \\
& = F(x, y, z) + 0.51 \cdot 2 + 0.91 \cdot 3 + 1 \cdot 5 \\
& = F(x, y, z) + F(2, 3, 5)
\end{align*}
$$


In [ ]:
# F(x, y) = 2 * x + 3 * y

def f(x, y):
    return 2 * x + 3 * y

# Visualisierung der Funktion für x und y im Bereich [0, 10]

x = np.linspace(-10, 10, 100)
y = np.linspace(-10, 10, 100)
X, Y = np.meshgrid(x, y)

Z = f(X, Y)

# Plot the Z surface using plotly
import plotly.graph_objects as go

fig = go.Figure(data=[go.Surface(z=Z, x=X, y=Y)])
fig.update_layout(title='Die Ebene definiert durch F(x, y) = 2x + 3y',
                  scene=dict(xaxis_title='x', yaxis_title='y', zaxis_title='f(X, Y)'))
# Add a marker at (0, 0, 0)
fig.add_trace(go.Scatter3d(x=[0], y=[0], z=[0], mode='markers', marker=dict(size=5, color='red')))
fig.show()
# Plot the Z surface using matplotlib